### PR TITLE
add evaluateOnNewDocument examples

### LINKED ISSUE #1201: Preloading a Script
This might not be an issue, more like a reference to how to mimic preload feature from nightmare/electron. 

If this looks good maybe we can include this in docs.
If this needs improvement, then please suggest.

If we want to change some default page variables before the page loads, we got no way of doing this in puppeteer. If we do, I don't see it in any docs or code.

For example, I might want to change navigator object properties and mock tests.

```js
// overwrite the `languages` property to use a custom getter
Object.defineProperty(navigator, 'languages', {
  get: function() {
    return ['en-US', 'en'];
  },
});

// overwrite the `plugins` property to use a custom getter
Object.defineProperty(navigator, 'plugins', {
  get: function() {
    // this just needs to have `length > 0`, but we could mock the plugins too
    return [1, 2, 3, 4, 5];
  },
});
```

It's easily doable with nightmare,

```js
var Nightmare = require('nightmare'),
  nightmare = Nightmare({
  webPreferences: {
        preload: 'alertMessage.js' // <-- this will get executed first, so we can change whatever object we want
      }
});
```

Note this is not addScriptTag or exposeFunction. exposeFunction will add a function, addScriptTag will do something on load. But, a preload script will load first, and make whatever changes we need to change.

I am testing this function right now, `evaluateOnNewDocument`. But defining a file which will load before every page load seems much more convenient.

Right now, I can do this.
```js
function preload() {
  // overwrite the `languages` property to use a custom getter
  Object.defineProperty(navigator, "languages", {
    get: function() {
      return ["en-US", "en", "bn", "github", "idontknowlanguage"];
    }
  });

  // overwrite the `plugins` property to use a custom getter
  Object.defineProperty(navigator, "plugins", {
    get: function() {
      // this just needs to have `length > 0`, but we could mock the plugins too
      return [1, 2, 3, 4, 5];
    }
  });
}

const page = await browser.newPage();
await page.evaluateOnNewDocument(preload);
```
Or like this in seperate file.
```js
// preload.js
module.exports = function(){
    // overwrite the `languages` property to use a custom getter
    Object.defineProperty(navigator, "languages", {
      get: function() {
        return ["en-US", "en", "bn", "c", "js"];
      }
    });

    // overwrite the `plugins` property to use a custom getter
    Object.defineProperty(navigator, "plugins", {
      get: function() {
        // this just needs to have `length > 0`, but we could mock the plugins too
        return [1, 2, 3, 4, 5];
      }
    });
}

// in out puppeteer
await page.evaluateOnNewDocument(require("./preload"));
```
Maybe we can include some preload scripts/list of files to preload in the `browser.newPage()`
for example, 

```js
browser.newPage({
 preload: "this.js"
})

// or something like this which will execute in order or async
browser.newPage({
 preload: ["this.js"],
 preloadAsync: true
})
```

### PR DESCRIPTION
**Pull request by entrptaher**
Fixes #1201

### CODE REVIEW

**Comment by entrptaher** [comment]
I signed it!

File: docs/api.md [hunk 1]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
 
 The function is invoked after the document was created but before any of its scripts were run. This is useful to amend JavaScript environment, e.g. to seed `Math.random`.
 
+An example of using preloading some functions from an external file would be following,
```
Comment (ebidel) [inline docs/api.md]: "An example of overriding the `navigator.languages` property before the page loads:"

File: docs/api.md [hunk 2]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
...
+
+```js
+// preload.js
+module.exports = function(){
```
Comment (ebidel) [inline docs/api.md]: space: `function() {`

File: docs/api.md [hunk 3]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
...
+    console.log("Adding some fake languages to navigator property");
+    
+    // overwrite the `languages` property to use a custom getter
+    Object.defineProperty(navigator, "languages", {
+      get: function() {
+        return ["en-US", "en", "c", "js"];
+      }
+    });
+}
```
Comment (ebidel) [inline docs/api.md]: `};`

File: docs/api.md [hunk 4]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
 
 The function is invoked after the document was created but before any of its scripts were run. This is useful to amend JavaScript environment, e.g. to seed `Math.random`.
 
+An example of using preloading some functions from an external file would be following,
+
+```js
+// preload.js
+module.exports = function(){
+    console.log("Adding some fake languages to navigator property");
```
Comment (ebidel) [inline docs/api.md]: IMO, the `console.log` doesn't add much value. Please remove.

File: docs/api.md [hunk 5]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
...
+    
+    // overwrite the `languages` property to use a custom getter
+    Object.defineProperty(navigator, "languages", {
+      get: function() {
+        return ["en-US", "en", "c", "js"];
+      }
+    });
+}
+
+// on the puppeteer script, assuming the preload.js file is in same folder of our script
```
Comment (ebidel) [inline docs/api.md]: just: `// In your puppeteer script.`

File: docs/api.md [hunk 6]
```diff
@@ -670,6 +670,25 @@ Adds a function which would be invoked in one of the following scenarios:
...
+await page.evaluateOnNewDocument(require("./preload"));
```
Comment (ebidel) [inline docs/api.md]: `"./preload.js"`

**Review by ebidel — CHANGES_REQUESTED** [review]
Just some nits and comment updates to make things more clear.

File: docs/api.md (hunk 6 above)
Comment (aslushnikov) [inline docs/api.md]: I'd drop the `require` part: it adds unnecessary indirection.
It also assumes commonJS modules, which is opinionated.

Other then that, great example!

Comment (entrptaher) [reply docs/api.md]: Is there any other easy way to preload an external file?

Comment (aslushnikov) [reply docs/api.md]: @entrptaher sure, you can use `fs` for this: `fs.readFileSync('filepath.js', 'utf8')`. Eventually the es6 imports would work as well.

Comment (entrptaher) [reply docs/api.md]: oh, that, I was wondering if you meant not to use any external file.
Let me fix these styles and things when I get time.

Comment (aslushnikov) [reply docs/api.md]: @entrptaher I suggested not to bother with an external file since it has nothing to do with the API itself.

**Comment by aslushnikov** [comment]
@entrptaher any updates on this?

File: docs/api.md [hunk 7]
```diff
@@ -670,6 +670,24 @@ Adds a function which would be invoked in one of the following scenarios:
 
 The function is invoked after the document was created but before any of its scripts were run. This is useful to amend JavaScript environment, e.g. to seed `Math.random`.
 
+An example of overriding the navigator.languages property before the page loads:
+
+```js
+// preload.js
+module.exports = function() {
```
Comment (aslushnikov) [inline docs/api.md]: if you use `readFile` method, then you shouldn't use `module.exports` part. You also don't need to define a wrapper function - the `readFile` returns a string.

So this file will simply contain the following:

```js
// preload.js  {
  get: function() {
    return ["en-US", "en", "c", "js"];
  };
});
```

**Review by aslushnikov — APPROVED** [review]

**Comment by aslushnikov** [comment]
@entrptaher thank you.

@ebidel how do you like this PR?

**Review by JoelEinbinder — APPROVED** [review]
I like it

**Review by ebidel — APPROVED** [review]

**Comment by entrptaher** [comment]
@aslushnikov do I have to do anything else? or it will merged later?

**Comment by aslushnikov** [comment]
@entrptaher do I get it right that you signed the CLA, but the googlebot is stuck?

**Comment by entrptaher** [comment]
@aslushnikov Seems so. I am sure I signed it.
Could it be that I used the web version to make pull request and thus it's stuck?

**Comment by ebidel** [comment]
@aslushnikov this is ready to merge

**Pull request merged**